src/mcs/montecarlo.jl

```diff
@@ -14,56 +16,73 @@ function print_nonempty(name, vector)
     end
 end
 
-function Base.show(io::IO, sim::Simulation{T}) where T <: AbstractSimulationData
-    println("Simulation{$T}")
+function Base.show(io::IO, sim_def::SimulationDef{T}) where T <: AbstractSimulationData
+    println("SimulationDef{$T}")
     
-    println("  trials: $(sim.trials)")
-    println("  current_trial: $(sim.current_trial)")
+    println("  trials: $(sim_def.trials)")
+    println("  current_trial: $(sim_def.current_trial)")
     
-    sim.current_trial > 0 && println("  current_data: $(sim.current_data)")
+    sim_def.current_trial > 0 && println("  current_data: $(sim_def.current_data)")
     
     println("  rvdict:")
-    for (key, value) in sim.rvdict
+    for (key, value) in sim_def.rvdict
         println("    $key: $(typeof(value))")
     end
 
-    print_nonempty("translist", sim.translist)
-    print_nonempty("savelist",  sim.savelist)
-    println("  nt_type: $(sim.nt_type)")
-    println("  $(length(sim.models)) models")
-    println("  $(length(sim.results)) results dicts")
+    print_nonempty("translist", sim_def.translist)
+    print_nonempty("savelist",  sim_def.savelist)
+    println("  nt_type: $(sim_def.nt_type)")
 
-    Base.show(io, sim.data)  # note: data::T
+    Base.show(io, sim_def.data)  # note: data::T
+end
+
+function Base.show(io::IO, sim_inst::SimulationInstance{T}) where T <: AbstractSimulationData
+    println("SimulationInstance{$T}")
+    
+    Base.show(io, sim_inst.sim_def)
+
+    println("  $(length(sim_inst.models)) models")
+    println("  $(length(sim_inst.results)) results dicts")    
 end
 
 function Base.show(obj::T) where T <: AbstractSimulationData
     nothing
 end
 
-# Store results for a single parameter
-function _store_param_results(m::Model, datum_key::Tuple{Symbol, Symbol}, trialnum::Int, results::Dict{Tuple, DataFrame})
+# Store results for a single parameter and return the dataframe for this particular
+# trial/scenario 
+function _store_param_results(m::Model, datum_key::Tuple{Symbol, Symbol}, trialnum::Int, scen_name::Union{Nothing, String}, results::Dict{Tuple, DataFrame})
     @debug "\nStoring trial results for $datum_key"
 
     (comp_name, datum_name) = datum_key
     dims = dimensions(m, comp_name, datum_name)
-            
+    has_scen = ! (scen_name === nothing)
+
     if length(dims) == 0        # scalar value
         value = m[comp_name, datum_name]
         # println("Scalar: $value")
 
         if haskey(results, datum_key)
             results_df = results[datum_key]
         else
-            results_df = DataFrame([typeof(value), Int], [datum_name, :trialnum], 0)
+            if has_scen
+                results_df = DataFrame([typeof(value), Int, String], [datum_name, :trialnum, :scen], 0)
+            else
+                results_df = DataFrame([typeof(value), Int], [datum_name, :trialnum], 0)
+            end
             results[datum_key] = results_df
         end
```

Comment:
This is probably a matter of taste, but in my experience redundancy always (eventually) creates more work because you have to change multiple things rather than one. I'd write this as follows, which slightly longer but without redundancy:
```
types = [typeof(value), Int]
names = [datum_name, :trialnum]
if has_scen
   push!(types, String)
   push!(names, :scen)
end
results_df = DataFrame(types, names, 0)
results[datum_key] = results_df
```